mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -0,0 +1,652 @@
+//===- MemoryAccessAnalysis.cpp - SYCL Memory Access Analysis -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h"
+#include "mlir/Analysis/DataFlow/IntegerRangeAnalysis.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/Debug.h"
+#include <utility>
+
+#define DEBUG_TYPE "memory-access-analysis"
+
+using namespace mlir;
+using namespace mlir::sycl;
+
+//===----------------------------------------------------------------------===//
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+/// Determine whether a value is an known integer value.
+static Optional<APInt> getConstIntegerValue(Value val, DataFlowSolver &solver) {
+  if (!val.getType().isIntOrIndex())
+    return std::nullopt;
+
+  auto *inferredRange =
+      solver.lookupState<dataflow::IntegerValueRangeLattice>(val);
+  if (!inferredRange || inferredRange->getValue().isUninitialized()) {
+    LLVM_DEBUG(llvm::dbgs() << "Solver not initialized correctly\n");
+    return std::nullopt;
+  }
+
+  const ConstantIntRanges &range = inferredRange->getValue().getValue();
+  LLVM_DEBUG(llvm::dbgs() << "range: " << range << "\n");
+
+  return range.getConstantValue();
+}
+
+/// Determine whether a value is equal to a given integer constant.
+static bool isEqualTo(Value val, int constant, DataFlowSolver &solver) {
+  Optional<APInt> optConstVal = getConstIntegerValue(val, solver);
+  if (!optConstVal.has_value())
+    return false;
+
+  APInt constVal = *optConstVal;
+  APInt c(constVal.getBitWidth(), constant, true /* signed */);
+  return (constVal == c);
+}
```

Comment:
```suggestion
static bool isEqualTo(Value val, uint64_t constant, DataFlowSolver &solver) {
  std::optional<APInt> optConstVal = getConstIntegerValue(val, solver);
  return optConstVal && (*optConstVal == constant);
}
```

Reply:
Oh, that's interesting... LGTM in that case. Thanks!

Reply:
It doesn't work when int is negative e.g., -1.